### PR TITLE
feat: user, wallet and pay commands with efficient error handling

### PR DESCRIPTION
**Pull request by eshanvaid**
**Key Changes:**

- Implemented `me`, `balance`, `set-username`, `default-wallet`, and `pay` commands
- Refactored the codebase to improve modularity and maintainability.
- Introduced a new error handling module with a unified error type hierarchy.

### CODE REVIEW

File: src/app/operations/globals.rs [hunk 1]
```diff
@@ -2,8 +2,15 @@ use crate::app::App;
...
-        let globals = self.client.globals().await?;
-        println!("{}", serde_json::to_string_pretty(&globals)?);
-        Ok(())
+        match self.client.globals().await {
+            Ok(globals) => {
+                println!("{}", serde_json::to_string_pretty(&globals)?);
+                Ok(())
+            }
+            Err(err) => {
+                println!("Error occurred while fetching globals");
```
Comment (bodymindarts) [inline src/app/operations/globals.rs:11]: if you are going to print in the error branch you should use `eprintln`.

But when using `anyhow` I think you can reduce the code to:
```
let globals = self.client.globals().await.context("Error while fetching globals")?
```

File: src/app/operations/intraledger_payment.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    pub async fn intraledger_payment(
+        &self,
+        username: String,
+        wallet: Wallet,
+        cents: Option<Decimal>,
+        sats: Option<Decimal>,
+        memo: Option<String>,
+    ) -> Result<()> {
+        if wallet == Wallet::Btc && sats.is_none() {
+            return Err(PaymentError::AmountNotSpecifiedBTC.into());
```
Comment (bodymindarts) [inline src/app/operations/intraledger_payment.rs:17]: Non optional arguments should be validated in the cli. Clap has good primitives for this that cover most needs.

File: src/app/operations/intraledger_payment.rs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                match self
+                    .client
+                    .intraleger_send_btc(username.clone(), sats, memo)
+                    .await
+                {
+                    Ok(()) => {
+                        println!("Successfully sent {} sats to username: {}", sats, username)
+                    }
+                    Err(err) => {
+                        println!("Error occurred while executing BTC intraledger payment ",);
```
Comment (bodymindarts) [inline src/app/operations/intraledger_payment.rs:36]: use `.context("...")?`

File: src/app/operations/intraledger_payment.rs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    ) -> Result<()> {
+        if wallet == Wallet::Btc && sats.is_none() {
+            return Err(PaymentError::AmountNotSpecifiedBTC.into());
+        }
+
+        if wallet == Wallet::Usd && cents.is_none() {
+            return Err(PaymentError::AmountNotSpecifiedUSD.into());
+        }
+
+        match wallet {
```
Comment (bodymindarts) [inline src/app/operations/intraledger_payment.rs:24]: I would add the amounts here and have a match with multiple cases. Then you don't need the early returns and you don't need the `expect` calls later on:

```
match (wallet, sats, cents) {
  (Wallet::Btc, None, _ ) => return Err(anyhow!("Amount not specified in Btc"))
  (Wallet::Btc, Some(sats), _) => <payment code>
  etc.
}
```

File: src/app/operations/me.rs [hunk 5]
```diff
@@ -2,11 +2,18 @@ use crate::app::App;
...
+        match self.client.me().await {
+            Ok(result) => {
+                println!(
+                    "{}",
+                    serde_json::to_string_pretty(&result).expect("Can't serialize json")
+                );
+                Ok(())
+            }
+            Err(err) => {
+                println!("Error occurred while fetching 'me' data");
```
Comment (bodymindarts) [inline src/app/operations/me.rs:14]: use `context("..")`

File: src/client/requests/globals.rs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+};
+
+impl GaloyClient {
+    pub async fn globals(&self) -> Result<QueryGlobalsGlobals, CliError> {
+        let variables = query_globals::Variables;
+
+        let response_body =
+            post_graphql::<QueryGlobals, _>(&self.graphql_client, &self.api, variables)
+                .await
+                .map_err(|_| ApiError::IssueGettingResponse)?;
```
Comment (bodymindarts) [inline src/client/requests/globals.rs:18]: I wouldn't eat the previous error - then you never find out what the issue might be. You can wrap it if you want but don't just drop it.

File: src/client/requests/intraledger_send_usd.rs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        memo: Option<String>,
+    ) -> Result<(), CliError> {
+        let me = self.me().await?;
+        let usd_wallet_id = me
+            .default_account
+            .wallets
+            .iter()
+            .find(|wallet| wallet.wallet_currency == WalletCurrency::USD)
+            .map(|wallet| &wallet.id)
+            .expect("Can't get USD wallet")
```
Comment (bodymindarts) [inline src/client/requests/intraledger_send_usd.rs:29]: don't `expect`

File: src/client/requests/intraledger_send_usd.rs [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+        let variables = intra_ledger_usd_payment_send::Variables { input };
+
+        let response_body = post_graphql::<IntraLedgerUsdPaymentSend, _>(
+            &self.graphql_client,
+            &self.api,
+            variables,
+        )
+        .await
+        .map_err(|_| ApiError::IssueGettingResponse)?;
```
Comment (bodymindarts) [inline src/client/requests/intraledger_send_usd.rs:48]: don't eat the error

File: src/client/requests/intraleger_send_btc.rs [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        memo: Option<String>,
+    ) -> Result<(), CliError> {
+        let me = self.me().await?;
+        let btc_wallet_id = me
+            .default_account
+            .wallets
+            .iter()
+            .find(|wallet| wallet.wallet_currency == WalletCurrency::BTC)
+            .map(|wallet| &wallet.id)
+            .expect("Can't get BTC wallet")
```
Comment (bodymindarts) [inline src/client/requests/intraleger_send_btc.rs:30]: `expect`

File: src/errors/payment_error.rs [hunk 10]
```diff
@@ -0,0 +1,9 @@
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum PaymentError {
+    #[error("For BTC wallet, use --sats to specify amount")]
```
Comment (bodymindarts) [inline src/errors/payment_error.rs:5]: This should be handled at CLI level.

File: src/client/requests/intraledger_send_usd.rs [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+};
+
+impl GaloyClient {
+    pub async fn intraleger_send_usd(
+        &self,
+        username: String,
+        amount: Decimal,
+        memo: Option<String>,
+    ) -> Result<(), CliError> {
+        let me = self.me().await?;
```
Comment (sandipndev) [inline src/client/requests/intraledger_send_usd.rs:22]: This is wrong - should be done in app layer.

File: src/app/operations/globals.rs (hunk 1 above)
Comment (eshanvaid) [reply src/app/operations/globals.rs]: The intent here was to produce better formatted error outputs. However, I observe that it doesn't make a significant difference. Replaced result matching with `.context` in 89c688ef6e37082aa02359d5e93497d1a0a19634 and 3b777a5797fa8e45d7b8a3a9326f6ccdfa8d7c88

File: src/app/operations/intraledger_payment.rs (hunk 2 above)
Comment (eshanvaid) [reply src/app/operations/intraledger_payment.rs]: Thank you for this, @bodymindarts. I fixed this in commit 7e229af6ad889eb9bc0791aa572444ce6d9ac221. It significantly streamlined the code.

File: src/app/operations/intraledger_payment.rs (hunk 3 above)
Comment (eshanvaid) [reply src/app/operations/intraledger_payment.rs]: Addressed

File: src/app/operations/me.rs (hunk 5 above)
Comment (eshanvaid) [reply src/app/operations/me.rs]: Addressed in latest commits

File: src/client/requests/globals.rs (hunk 6 above)
Comment (eshanvaid) [reply src/client/requests/globals.rs]: Addressed this in a106ff98bbe1579b58865e8ac62717b0fafca644

File: src/client/requests/intraledger_send_usd.rs (hunk 8 above)
Comment (eshanvaid) [reply src/client/requests/intraledger_send_usd.rs]: Addressed in a106ff98bbe1579b58865e8ac62717b0fafca644

File: src/errors/payment_error.rs (hunk 10 above)
Comment (eshanvaid) [reply src/errors/payment_error.rs]: 7e229af6ad889eb9bc0791aa572444ce6d9ac221

File: src/client/requests/fetch_balance.rs [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+};
+
+impl GaloyClient {
+    pub async fn fetch_balance(
+        &self,
+        btc: bool,
+        usd: bool,
+        wallet_ids: Vec<String>,
+    ) -> anyhow::Result<Vec<WalletBalance>> {
+        let me = self.me().await?;
```
Comment (sandipndev) [inline src/client/requests/fetch_balance.rs:15]: same issue - can you have this in app layer.

**Review by sandipndev — APPROVED** [review]

**Pull request merged**